src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+        /// <param name="unspentOutputs">The unspent outputs.</param>
+        private void CheckCoinstakeAgeRequirement(ChainedHeader chainedHeader, UnspentOutputs unspentOutputs)
+        {
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            var options = (PosConsensusOptions)this.PosParent.Network.Consensus.Options;
+            int targetDepth = options.GetStakeMinConfirmations(prevChainedHeader.Height + 1, this.PosParent.Network) - 1;
+
+            int actualDepth = prevChainedHeader.Height - (int)unspentOutputs.Height;
+            if (actualDepth < targetDepth)
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_STAKE_DEPTH]");
+            ConsensusErrors.InvalidStakeDepth.Throw();
+        }
+
+        /// <see cref="IStakeValidator.VerifySignature"/>
+        private void CheckSignature(ProvenBlockHeader header, UnspentOutputs unspentOutputs)
+        {
+            if (this.stakeValidator.VerifySignature(unspentOutputs, header.Coinstake, 0, ScriptVerify.None))
+                return;
+            
+            this.Logger.LogTrace("(-)[BAD_SIGNATURE]");
+            ConsensusErrors.CoinstakeVerifySignatureFailed.Throw();
+        }
+
+        /// <see cref="IStakeValidator.CheckStakeKernelHash"/>
+        private void CheckStakeKernelHash(PosRuleContext context, UnspentOutputs stakingCoins, ProvenBlockHeader header, ChainedHeader chainedHeader)
+        {
+            OutPoint prevOut = this.GetPreviousOut(header);
+            uint transactionTime = header.Coinstake.Time;
+
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            BlockStake prevBlockStake = this.PosParent.StakeChain.Get(prevChainedHeader.HashBlock);
+            if (prevBlockStake == null)
+            {
+                this.Logger.LogTrace("(-)[BAD_PREV_STAKE]");
+                ConsensusErrors.PrevStakeNull.Throw();
+            }
+
+            uint headerBits = chainedHeader.Header.Bits.ToCompact();
+
+            this.stakeValidator.CheckStakeKernelHash(context, headerBits, prevBlockStake, stakingCoins, prevOut, transactionTime);
+        }
+
+        /// <summary>
+        /// Check that coinstake tx is in the merkle tree using the merkle proof.
+        /// </summary>
+        /// <param name="header">The header.</param>
+        private void CheckCoinstakeMerkleProof(ProvenBlockHeader header)
+        {
+            if (header.MerkleProof.Check(header.HashMerkleRoot))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_MERKLE_ROOT]");
+            ConsensusErrors.BadMerkleRoot.Throw();
+        }
+
+        /// <summary>
+        /// Verifies header signature with the key from coinstake kernel.
+        /// </summary>
+        /// <param name="header">The header.</param>
+        /// <param name="stakingCoins">The staking coins.</param>
+        private void CheckHeaderSignatureWithConinstakeKernel(ProvenBlockHeader header, UnspentOutputs stakingCoins)
```

Comment:
What's the difference with `CheckSignature()`?

Reply:
`CheckSignature()` checks the script signature of the coinstake transaction input, while `CheckHeaderSignatureWithConinstakeKernel()` verifies header signature with the key from coinstake kernel. I think it is better explained in PH research article.

Reply:
@fenix2222 @dangershony  Typo? CheckHeaderSignatureWith**Coinstake**Kernel not **Conin**

Reply:
Ah yeah we need to verify the block signature as well ofcourse.